File: sdk/java/src/main/java/io/forma/Invocation.java
package io.forma;

import java.util.Collections;
import java.util.Map;

/**
 * One action invocation from the sidecar — the wire form of the engine's
 * ExecuteParams (docs/runtimes/04-forma-sidecar.md §4.2).
 *
 * @param module     owning module name
 * @param entity     entity name
 * @param action     action name
 * @param resourceId identifier of the resource being acted upon
 * @param resource   current entity record
 * @param params     action parameters from the request body
 * @param tenantId   tenant context
 * @param userId     user context
 */
public record Invocation(
        String module,
        String entity,
        String action,
        String resourceId,
        Map<String, Object> resource,
        Map<String, Object> params,
        String tenantId,
        String userId) {

    @SuppressWarnings("unchecked")
    static Invocation fromRequest(String module, String entity, String action, Map<String, Object> body) {
        return new Invocation(
                module,
                entity,
                action,
                (String) body.getOrDefault("resource_id", ""),
                (Map<String, Object>) body.getOrDefault("resource", Collections.emptyMap()),
                (Map<String, Object>) body.getOrDefault("params", Collections.emptyMap()),
                (String) body.getOrDefault("tenant_id", ""),
                (String) body.getOrDefault("user_id", ""));
    }
}

File: sdk/dotnet/src/FormSpecLib/Invocation.cs
namespace FormSpec;

/// <summary>
/// One action invocation from the sidecar — the wire form of the engine's
/// ExecuteParams (docs/runtimes/04-formspec-sidecar.md §4.2).
/// </summary>
public sealed record Invocation(
    string Module,
    string Entity,
    string Action,
    string ResourceId,
    Dictionary<string, object?> Resource,
    Dictionary<string, object?> Params,
    string UserId)
{
    internal static Invocation FromRequest(
        string module, string entity, string action,
        Dictionary<string, object?> body)
    {
        return new Invocation(
            module,
            entity,
            action,
            (string)body.GetValueOrDefault("resource_id", "")!,
            (Dictionary<string, object?>)body.GetValueOrDefault("resource", new())!,
            (Dictionary<string, object?>)body.GetValueOrDefault("params", new())!,
            (string)body.GetValueOrDefault("user_id", "")!);
    }
}
